### PR TITLE
Upload size help text. ref 5652

### PR DESCRIPTION
**Pull request by OaklandPeters**

### CODE REVIEW

File: cropduster/static/cropduster/js/upload.js [hunk 1]
```diff
@@ -127,6 +127,38 @@
...
+                return;
+            }
+            var min_w = autoSize.min_w || autoSize.w || 0;
+            var min_h = autoSize.min_h || autoSize.h || 0;
+            minSize[0] = Math.max(minSize[0], min_w);
+            minSize[1] = Math.max(minSize[1], min_h);
+        });
+        return minSize;
+    };
+    var largest = function largest(array, initial, keyfunc) {
```
Comment (fdintino) [inline cropduster/static/cropduster/js/upload.js]: You could probably just replace this with `Math.max.apply(null, array) || 1`

File: cropduster/static/cropduster/js/upload.js [hunk 2]
```diff
@@ -127,6 +127,38 @@
...
+            return (keyfunc(left) >= keyfunc(right)) ? left: right;
+        });
+    }
+    // Fill in 'Min Size' help text
+    $(document).ready(function(){
+        var sizes = JSON.parse($('#id_crop-sizes').val());
+        var minSizes = sizes.map(calcMinSize);
+        var largest_width = largest(minSizes.map(function getWidth(sz){ return sz[0] }), 1);
+        var largest_height = largest(minSizes.map(function getHeight(sz){ return sz[1] }), 1);
+        $("#upload-help #min-width").html(String(largest_width));
```
Comment (fdintino) [inline cropduster/static/cropduster/js/upload.js]:
```javascript
var isStandalone = $('body').is('.cropduster-standalone');
if (isStandalone || !sizes) {
    return;
}
var largest = [
    Math.max.apply(null, minSizes.map(function(s) { return s[0]; })),
    Math.max.apply(null, minSizes.map(function(s) { return s[1]; }))
];
$('.row.image').append(
    $('<div id="upload-help">Min. size: ' + largest.join(' x ') + '</div>'));
```

Comment (fdintino) [inline cropduster/static/cropduster/js/upload.js]: With the above you can remove the need to change anything in upload.html

Comment (OaklandPeters) [inline cropduster/static/cropduster/js/upload.js]: True - but that has a significant drawback. By removing the placeholder tags from the template, the help-related content becomes 'invisible' to a developer working at the level of the HTML template/CSS.

TL;DR Inserting content via JS, without at least a placeholder tag, makes the life of later developers more difficult.

Comment (fdintino) [inline cropduster/static/cropduster/js/upload.js]: Does it? I think it's a wash. Either you have logic spread out to multiple places but the behavior is clearer when looking at one thing individually, or it's all contained in a single place. I come down on the side of keeping things together. Also the way you had it, it would show up as a dumb "Min. size: x" when it didn't apply (for instance, in CKEditor mode).

Comment (fdintino) [inline cropduster/static/cropduster/js/upload.js]: `"upload-help"` is not the clearest name anyway...

Comment (OaklandPeters) [inline cropduster/static/cropduster/js/upload.js]: A compromise:

`<div id="upload-help"></div>`

and
```
var isStandalone = $('body').is('.cropduster-standalone');
if (isStandalone || !sizes) {
    return;
}
var largest = [
    Math.max.apply(null, minSizes.map(function(s) { return s[0]; })) || 1,
    Math.max.apply(null, minSizes.map(function(s) { return s[1]; })) || 1
];
$('#upload-help').html(Min. size: ' + largest.join(' x ')));
```

Comment (OaklandPeters) [inline cropduster/static/cropduster/js/upload.js]: `upload-help` --> `upload-min-size-help`?

Comment (fdintino) [inline cropduster/static/cropduster/js/upload.js]: Sure. I think the concern is really only valid in a world where developers don't use Web Inspectors, but if you feel strongly about it the above is fine.
